### PR TITLE
#124141 preliminaries

### PR DESCRIPTION
**Pull request by nnethercote**
Preliminary changes required to start removing `Nonterminal` (https://github.com/rust-lang/rust/pull/124141).

r? @petrochenkov

### CODE REVIEW

File: compiler/rustc_ast/src/token.rs [hunk 1]
```diff
@@ -42,11 +42,87 @@ pub enum BinOpToken {
...
+    // From the expansion of a metavariable in a declarative macro.
+    MetaVar(MetaVarKind),
+
+    // Converted from `proc_macro::Delimiter` in
+    // `proc_macro::Delimiter::to_internal`, i.e. returned by a proc macro.
+    ProcMacro,
+
+    // Converted from `TokenKind::Interpolated` in
+    // `TokenStream::flatten_token`. Treated similarly to `ProcMacro`.
+    FlattenToken,
```
Comment (petrochenkov) [inline compiler/rustc_ast/src/token.rs:57]: Is there any observable difference between `ProcMacro` and `FlattenToken`?
(It's possible that I already asked this on the older PRs.)

File: compiler/rustc_ast/src/token.rs [hunk 2]
```diff
@@ -42,11 +42,87 @@ pub enum BinOpToken {
...
+            Ty => sym::ty,
+            Ident => sym::ident,
+            Lifetime => sym::lifetime,
+            Literal => sym::literal,
+            Meta => sym::meta,
+            Path => sym::path,
+            Vis => sym::vis,
+            TT => sym::tt,
+        };
+        write!(f, "{}", sym)
```
Comment (petrochenkov) [inline compiler/rustc_ast/src/token.rs:118]:
```suggestion
        write!(f, "{sym}")
```

File: compiler/rustc_ast/src/token.rs [hunk 3]
```diff
@@ -59,7 +135,33 @@ pub enum Delimiter {
...
+    Invisible(InvisibleOrigin),
+}
+
+impl Delimiter {
+    // Should the parser skip these delimiters? Only happens for certain kinds
+    // of invisible delimiters. Ideally this function will eventually disappear
+    // and no invisible delimiters will be skipped.
+    #[inline]
+    pub fn skip(&self) -> bool {
+        use InvisibleOrigin::*;
```
Comment (petrochenkov) [inline compiler/rustc_ast/src/token.rs:147]: Could you avoid glob importing enums?

I've seen you doing it as a part of cleanups to various compiler crates, but I don't think it's an improvement (only in exceptional cases maybe, but nothing in this PR looks like that).

**Comment by petrochenkov** [comment]
Quite a few perf regressions, do you know what is the reason? Are they avoidable?
@rustbot author

File: compiler/rustc_ast/src/token.rs (hunk 1 above)
Comment (nnethercote) [reply compiler/rustc_ast/src/token.rs:57]: There is no difference. The final commit in #124141 removes `FlattenToken` and flattening in general.

File: compiler/rustc_ast/src/token.rs (hunk 3 above)
Comment (nnethercote) [reply compiler/rustc_ast/src/token.rs:147]: Ok, I'll remove this one.

**Comment by nnethercote** [comment]
The regressions are smaller when I benchmark locally, so it's hard to tell exactly what's happening. But I suspect there are two possible causes:
- We used to just skip all invisible delimiters, but now we only skip some, which involves an extra check. This is most important in `TokenCursor::inlined_next` which is very hot. This is impossible to avoid.
- The extra cases added to `can_begin_*`/`may_be_*` will make those functions slightly slower. Once `Nonterminal` is removed some of the old cases will be removed, so any effect here should eventually disappear.

The regressions are mostly in `incr-unchanged` runs, where parsing/expansion performance changes are magnified. I can't see how to avoid these for now.

**Comment by petrochenkov** [comment]
r=me after removing the remaining glob imports (https://github.com/rust-lang/rust/pull/132629#discussion_r1835502125).
@rustbot author

**Comment by nnethercote** [comment]
I addressed all the comments.

@bors r=petrochenkov

**Comment by matthiaskrgr** [comment]
```
Run src/ci/scripts/setup-upstream-remote.sh
Configuring https://github.com/rust-lang/rust remote as upstream.
fatal: unable to access 'https://github.com/rust-lang/rust/': Failed to connect to github.com port 443: Connection timed out
```
@bors retry Failed to connect to github.com port 443: Connection timed out

**Pull request merged**